src/logic/import/voc/VOCImporter.ts

```diff
@@ -0,0 +1,98 @@
+import {ImageData, LabelName, LabelRect} from '../../../store/labels/types';
+import {LabelUtil} from "../../../utils/LabelUtil";
+import {AnnotationImporter} from '../AnnotationImporter';
+import {LabelsSelector} from '../../../store/selectors/LabelsSelector';
+
+export type FileParseResult = {
+    filename: string,
+    labeledBoxes: LabelRect[]
+};
+export type VOCImportResult = {
+    labelNames: LabelName[],
+    fileParseResults: FileParseResult[],
+};
+
+export class VOCImporter extends AnnotationImporter {
+    public static requiredKeys = ['annotation', 'filename', 'categories'];
+    private labelNames: Map<string, LabelName>;
+
+    public import(
+        filesData: File[],
+        onSuccess: (imagesData: ImageData[], labelNames: LabelName[]) => any,
+        onFailure: (error?:Error) => any
+    ): void {
+        try {
+            const inputImagesData: Map<string, ImageData> = VOCImporter.mapImageData();
+            this.labelNames = new Map<string, LabelName>();
+
+            Promise.all(this.loadAndParseFiles(filesData)).then(results => {
+                for (const result of results) {
+                    if (inputImagesData[result.filename]) {
+                        inputImagesData[result.filename].labelRects = result.labeledBoxes;
+                    }
+                }
+
+                onSuccess(
+                    Array.from(Object.values(inputImagesData)),
+                    Array.from(this.labelNames.values())
+                );
+            }).catch((error: Error) => onFailure(error));
+        } catch (error) {
+            onFailure(error as Error)
+        }
+    }
+
+    private loadAndParseFiles(files: File[]): Promise<FileParseResult>[] {
+        const parser = new DOMParser();
+        return files.map(fileData => fileData.text().then(text => 
+            this.parseDocumentIntoImageData(parser.parseFromString(text, 'application/xml'))
+        ));
+    }
+
+    private parseDocumentIntoImageData(document: Document): FileParseResult {
+        const root = document.getElementsByTagName('annotation')[0];
+        const filename = root.getElementsByTagName('filename')[0].textContent;
+        
+        const labeledBoxes: LabelRect[] = this.parseAnnotationsFromFileString(document);
+
+        return {
+            filename,
+            labeledBoxes,
+        };
+    }
+
+    private parseAnnotationsFromFileString(document: Document): LabelRect[] {
```

Comment:
This seems like a perfect method to write tests, but because you use `this.labelNames` in the method you will need to mock the whole class. We can of course do that. But I'd try to refactor that method and convert it into util/static method, which takes some `labelNames` and return new `labelNames`, and then you update the state of the clas outside of that util. How about that?

Reply:
Yeah, I didn't love the pattern as well for that reason. I was leaning towards a reducer approach to keep it more functional but I know that also makes it harder to read for those that don't know that style. Updated it to be like that. Let me know what you think